_posts/2018-07-25-dota-macos-performance.md

```diff
@@ -0,0 +1,64 @@
+
+Note that technically, "Immediate" recording has less CPU overhead, as well as more explicit threading: no surprises at submit time. We consider it to be the superior mode, however the application needs to make sure to not hold on the recorded command buffers for too long, and submit more often than once per frame. Otherwise, the latency kicks in, and Dota being programmed to allow at most a single frame of latency is able to produce less frames.
+
+Another interesting observation is that 2/3 of the immediate command recording time is spent inside the driver, according to the instrumented profile. This gives us an upper bound for Vulkan Portability overhead on Metal to be about 50%, although that includes some OS and Metal runtime bits as well.
+
+### Results
+
+| Library         | % of Main thread | fps - simple | fps - complex/low | fps variability/low | fps - complex/high | fps variability/high |
+| --------------- | ---------------- | ------------ | ------------- | --------------- | -------- | ---- |
+| gfx / immediate | 35%              | 90 fps       | 37.0 fps      | 5.7             | 31.0 fps | 4.5  |
+| gfx / deferred  | 15%              | 110 fps      | 44.4 fps      | 5.7             | 36.2 fps | 5.3  |
+| MoltenVK        | 21%              | 100 fps      | 41.7 fps      | 6.4             | 35.2 fps | 5.7  |
+| OpenGL          | ?                | 105-110 fps  | 45.6 fps      | 5.7             | 33.9 fps | 6.8  |
+
+The first metric shows how much time of the main thread is spent inside the portability library, compared to the total execution time (which is the actual time minus all the sleeping). It was measured with Instruments on a simple scene ("fps - simple" contains the frame rate on it). Note that the submission is done on a separate thread by Dota, so that time isn't taken into account here. Interestingly, the less time we spend on the main thread the faster our frame rate ends up being. Or, in other words, it's the race of who gets to the submission first :)
+
+The "fps - complex" and "fps variability" are from the `Source2Bench.csv` produced by timing the [dota2-pts-1971360796.dem](http://www.phoronix-test-suite.com/benchmark-files/dota2-pts-1971360796.dem.tar.bz2) demo in range 40000 .. 50000. The "/low" group refers to `-autoconfig_level 1 -low` command line, while the "/high" group is for `-autoconfig_level 3 -high`. We followed the instructions provided by [GamingOnLinux](https://www.gamingonlinux.com/articles/want-to-benchmark-dota-2-on-linux-heres-how-to-do-it.7435). To be fair, we were only testing on the simple scene until this point, so more discoveries are possible once we start profiling this complex demo.
```

Comment:
'in range 40000 .. 50000', it's not obvious what range you are referencing here but it probably should be 'in the range 40000 .. 50000'